### PR TITLE
Support translation via ChatGPT

### PR DESCRIPTION
**Pull request by taahamahdi**

### CODE REVIEW

File: src/chat_interface/chatgpt.ts [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+
+            const responseText = response.choices[0].message.content;
+            if (!responseText) {
+                return "";
+            }
+
+            this.history.push({ role: Role.System, content: responseText });
+            return responseText;
+        } catch (err) {
+            console.error(err);
```
Comment (Brainicism) [inline src/chat_interface/chatgpt.ts:54]: nit: itll be easier to debug with more context on where this error is being thrown from. We might be lucky and the error might imply it, but not guaranteed

File: src/chat_interface/chatgpt.ts [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+        }
+    }
+
+    resetChatHistory(): void {
+        this.history = [];
+    }
+
+    rollbackLastMessage(): void {
+        if (this.history[this.history.length - 1].role === Role.System) {
+            this.history.pop();
```
Comment (Brainicism) [inline src/chat_interface/chatgpt.ts:65]: Why do we pop twice for system prompt? Should document this

File: src/chat_interface/chatgpt.ts [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+            this.history.pop();
+        } else if (this.history[this.history.length - 1].role === Role.User) {
+            this.history.pop();
+        }
+    }
+
+    invalidTranslation(): void {
+        this.history.push({
```
Comment (Brainicism) [inline src/chat_interface/chatgpt.ts:73]: Are these prompts the same across both?

File: src/chat_interface/chatgpt.ts (hunk 2 above)
Comment (taahamahdi) [reply src/chat_interface/chatgpt.ts:65]: Good call, I'm popping twice to reach the last successful state. This means the history should revert back to the last successful response from the model. When the model was last to respond, we need to rollback both its response and its trigger, i.e. the user

File: src/chat_interface/chatgpt.ts (hunk 3 above)
Comment (taahamahdi) [reply src/chat_interface/chatgpt.ts:73]: Yeah, do you think they should be moved to the base class?

Comment (Brainicism) [reply src/chat_interface/chatgpt.ts:73]: yes

Comment (taahamahdi) [reply src/chat_interface/chatgpt.ts:73]: Looks to be difficult since both derived classes use a different type for `this.history`

Comment (Brainicism) [reply src/chat_interface/chatgpt.ts:73]: u can move just the strings

Comment (taahamahdi) [reply src/chat_interface/chatgpt.ts:73]: Good idea, thanks

**Review by Brainicism — APPROVED** [review]

**Pull request merged**